File: solutions/clojure/difference-of-squares/1/src/difference_of_squares.clj
(ns difference-of-squares)


(defn square [x] (* x x))

(defn sum-of-squares [n]
  (quot (* n (inc n) (inc (* 2 n))) 6))

(defn sum-n [n] (quot (* n (inc n)) 2))

(defn square-of-sum [n]
  (square (sum-n n)))

(defn difference [n]
  (- (square-of-sum n) (sum-of-squares n)))

File: solutions/python/difference-of-squares/1/difference_of_squares.py
"""Calculate differences between sums and squares."""


def square_of_sum(number):
    """Return the square of the sum of the first natural numbers."""
    total = number * (number + 1) // 2
    return total * total


def sum_of_squares(number):
    """Return the sum of the squares of the first natural numbers."""
    return number * (number + 1) * (2 * number + 1) // 6


def difference_of_squares(number):
    """Return the difference between square of sum and sum of squares."""
    return square_of_sum(number) - sum_of_squares(number)
